src/off_chain/historical_data/cow_endpoint_surplus.py

```diff
@@ -0,0 +1,295 @@
+    """
+
+    def get_settlement_hashes(self, start_block: int, end_block: int) -> List[str]:
+        etherscan_url = f"https://api.etherscan.io/api?module=account&action=txlist&address=0x9008D19f58AAbD9eD0D60971565AA8510560ab41&startblock={start_block}&endblock={end_block}&sort=desc&apikey={directory.ETHERSCAN_KEY}"
+        # all "result" go into results (based on API return value names from docs)
+        try:
+            settlements = json.loads(
+                (
+                    requests.get(
+                        etherscan_url,
+                        headers=header,
+                        timeout=1000000,
+                    )
+                ).text
+            )["result"]
+            settlement_hashes_list = []
+            for settlement in settlements:
+                settlement_hashes_list.append(settlement["hash"])
+        except ValueError:
+            self.logger.critical("etherscan error.")
+
+        return settlement_hashes_list
+
+    """
+    This function uses a list of tx hashes to fetch and assemble competition data
+    for each of the tx hashes and returns it to get_surplus_by_input for further surplus calculation.
+    """
+
+    def get_solver_competition_data(
+        self, settlement_hashes_list: List[str]
+    ) -> List[Dict[str, Any]]:
+        solver_competition_data = []
+        for tx_hash in settlement_hashes_list:
+            endpoint_url = f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{tx_hash}"
+            json_competition_data = requests.get(
+                endpoint_url,
+                headers=header,
+                timeout=1000000,
+            )
+            if json_competition_data.status_code == 200:
+                solver_competition_data.append(json.loads(json_competition_data.text))
+
+        return solver_competition_data
+
+    """
+    This function goes through each order that the winning solution executed and finds non-winning
+    solutions that executed the same rder and calculates surplus difference between that pair 
+    (winning and non-winning solution). 
+    Difference conversions to ETH and % deviations from traded amount have been made to check for flagging orders.
+    """
+
+    def get_order_surplus(self, competition_data: Dict[str, Any]) -> None:
+        winning_solver = competition_data["solutions"][-1]["solver"]
+        winning_orders = competition_data["solutions"][-1]["orders"]
+
+        for individual_win_order in winning_orders:
+            solver_dict[winning_solver][0] += 1
+            self.total_orders += 1
+            individual_win_order_id = individual_win_order["id"]
+            order_data_url = (
+                f"https://api.cow.fi/mainnet/api/v1/orders/{individual_win_order_id}"
+            )
+            json_order = requests.get(
+                order_data_url,
+                headers=header,
+                timeout=1000000,
+            )
+            if json_order.status_code == 200:
```

Comment:
If the API fails here, maybe we can add a logging message that says something like  "failed to check order uid=..."